projects/ADuCM3029_demo_cn0428_cn0429/src/adi_m355_gas_sensor.cpp

```diff
@@ -0,0 +1,824 @@
+*
+*IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, INTELLECTUAL PROPERTY
+*RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+*BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+*THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************
+**/
+
+#include "adi_m355_gas_sensor.h"
+#include "adi_m355_gas_sensor_config.h"
+#include "base_sensor/adi_sensor_errors.h"
+#include "ADuCM3029_demo_cn0428_cn0429.h"
+
+namespace adi_sensor_swpack
+{
+
+/*
+ * Open M355 gas sensor
+ */
+SENSOR_RESULT M355_GAS::open()
+{
+	pADI_GPIO0->DS |= (1 << 4) | (1 << 5);
+	this->m_i2c_address = ADI_CFG_I2C_DEFAULT_ADDR;
+	return(this->InitI2C());
+}
+
+/*
+ * Open M355 gas sensor with selected address
+ */
+SENSOR_RESULT M355_GAS::openWithAddr(uint8_t sensor_address)
+{
+	pADI_GPIO0->DS |= (1 << 4) | (1 << 5);
+	this->m_i2c_address = sensor_address;
+	return(this->InitI2C());
+}
+/*
+ * Starts measurement
+ */
+SENSOR_RESULT   M355_GAS::start()
+{
+	return(SENSOR_ERROR_NONE);
+}
+
+/*
+ * Stops measurement
+ */
+SENSOR_RESULT   M355_GAS::stop()
+{
+	return(SENSOR_ERROR_NONE);
+}
+
+/*
+ * Close device
+ */
+SENSOR_RESULT   M355_GAS::close()
+{
+	ADI_I2C_RESULT eI2cResult;
+
+	if((eI2cResult = adi_i2c_Close(m_i2c_handle)) == ADI_I2C_SUCCESS)
+		return SENSOR_ERROR_NONE;
+	else
+		return SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult);
+}
+
+/**
+ * @brief  Returns the temperature value
+ *
+ * @return SENSOR_RESULT
+ *
+ * @details
+ */
+SENSOR_RESULT M355_GAS::InitI2C()
+{
+	ADI_I2C_RESULT eI2cResult = ADI_I2C_SUCCESS;
+
+	eI2cResult = adi_i2c_Open(ADI_CFG_I2C_DEV_NUM, &m_I2C_memory,
+				  ADI_I2C_MEMORY_SIZE, &m_i2c_handle);
+	if(eI2cResult == ADI_I2C_SUCCESS) {
+		if((eI2cResult = adi_i2c_Reset(m_i2c_handle)) == ADI_I2C_SUCCESS) {
+			if((eI2cResult = adi_i2c_SetBitRate(m_i2c_handle,
+							    ADI_CFG_I2C_BIT_RATE))  == ADI_I2C_SUCCESS) {
+				return (SENSOR_ERROR_NONE);
+			}
+		}
+	}
+
+	return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
+}
+
+
+/**
+ * @brief  reads or writes data to/from the gas sensor
+ *
+ * @param  uint8_t RW, uint8_t RegAddr, uint8_t *pData, uint16_t size
+ *
+ * @return returns result of I2C communication
+ *
+ * @details
+ */
+SENSOR_RESULT M355_GAS::I2CReadWrite(uint8_t RW, uint8_t RegAddr,
+				     uint8_t *pData, uint16_t size)
+{
+	ADI_I2C_TRANSACTION sTransfer;
+	ADI_I2C_RESULT      eI2cResult = ADI_I2C_SUCCESS;
+	uint32_t            nHwErrors;
+	uint8_t             aPrologueData[5u];
+
+	/* Set slave address for ADI M355 gas sensor */
+	eI2cResult = adi_i2c_SetSlaveAddress(m_i2c_handle, this->m_i2c_address);
+
+	if (eI2cResult == ADI_I2C_SUCCESS) {
```

Comment:
it's a good idea to reduce the levels of indentation;
typically what we do [or try to do] is:
```
   if (eI2cResult != ADI_I2C_SUCCESS)
        return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
```

the `if (RW == READ) ` and `if(RW == WRITE)` can be un-indented one level if doing this